File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	READY = iota
	VISITED
	LOOP
	ISOLATED
)

func noLoopCnt(nextStud []int, n int) int {
	states := make([]int, n+1)
	for i := 1; i <= n; i++ {
		states[i] = READY
	}

	for i := 1; i <= n; i++ {
		if states[i] != READY {
			continue
		}

		// mark visited
		j := i
		for states[j] == READY {
			states[j] = VISITED
			j = nextStud[j]
		}

		// mark loop
		k := j
		for states[k] == VISITED {
			states[k] = LOOP
			k = nextStud[k]
		}

		// mark isolated
		s := i
		for states[s] == VISITED {
			states[s] = ISOLATED
			s = nextStud[s]
		}
	}

	cnt := 0
	for _, state := range states {
		if state == ISOLATED {
			cnt++
		}
	}

	return cnt
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var tc int
	fmt.Fscanf(reader, "%d\n", &tc)

	for t := 0; t < tc; t++ {
		var n int
		fmt.Fscanf(reader, "%d\n", &n)

		line, _ := reader.ReadString('\n')
		lineSpl := strings.Split(
			strings.TrimSpace(line),
			" ",
		)

		nextStud := make([]int, n+1)
		for i := 1; i <= n; i++ {
			nextStud[i], _ = strconv.Atoi(lineSpl[i-1])
		}

		cnt := noLoopCnt(nextStud, n)
		writer.WriteString(fmt.Sprintf("%d\n", cnt))
	}
}

File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline
output = sys.stdout.write
flush = sys.stdout.flush

from typing import *
from enum import Enum, auto
# from pprint import pprint

class State(Enum):
    READY = auto()
    VISITED = auto()
    LOOP = auto()
    ISOLATED = auto()



def main(nextStud: List[int], n: int) -> int:
    states = [State.READY for _ in range(n+1)]
    cnt = 0

    for i in range(1, n+1):
        # pprint(states)

        if states[i] is not State.READY:
            continue

        # mark visited
        j = i 
        while states[j] == State.READY:
            states[j] = State.VISITED
            j = nextStud[j]

        # mark loop
        k = j
        while states[k] != State.LOOP and states[k] != State.ISOLATED:
            states[k] = State.LOOP
            k = nextStud[k]

        # mark isolated
        s = i
        while states[s] != State.LOOP and states[s] != State.ISOLATED:
            states[s] = State.ISOLATED
            cnt += 1
            s = nextStud[s]

    return cnt


if __name__ == "__main__":
    n_tc = int(input())
    for _ in range(n_tc):
        n = int(input())
        vals = [0] + [int(x) for x in input().split()]
        result = main(vals, n)
        output(f"{result}\n")

    flush()